# Assignment Question 4

In [9]:
!wget http://www.amlbook.com/data/zip/features.train
!wget http://www.amlbook.com/data/zip/features.test

--2021-10-07 16:00:22--  http://www.amlbook.com/data/zip/features.train
Resolving www.amlbook.com (www.amlbook.com)... 50.62.137.42
Connecting to www.amlbook.com (www.amlbook.com)|50.62.137.42|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 357259 (349K)
Saving to: ‘features.train.3’

features.train.3    100%[===================>] 348.89K   842KB/s    in 0.4s    

2021-10-07 16:00:22 (842 KB/s) - ‘features.train.3’ saved [357259/357259]

--2021-10-07 16:00:23--  http://www.amlbook.com/data/zip/features.test
Resolving www.amlbook.com (www.amlbook.com)... 50.62.137.42
Connecting to www.amlbook.com (www.amlbook.com)|50.62.137.42|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 98343 (96K)
Saving to: ‘features.test.3’

features.test.3     100%[===================>]  96.04K   465KB/s    in 0.2s    

2021-10-07 16:00:23 (465 KB/s) - ‘features.test.3’ saved [98343/98343]



In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [11]:
header_list = ["result","feature1","feature2"]
data=pd.read_csv("/content/features.train",sep='  ',names=header_list)
test_data=pd.read_csv("/content/features.test",sep='  ',names=header_list)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  This is separate from the ipykernel package so we can avoid doing imports until


In [12]:
test_data.head()

result  feature1  feature2
0     9.0  0.272178 -4.847937
1     6.0  0.265133 -5.102000
2     3.0  0.335926 -2.921562
3     6.0  0.264850 -4.156625
4     6.0  0.345338 -6.718438

In [13]:
data.head()

result  feature1  feature2
0     6.0  0.341092 -4.528937
1     5.0  0.444131 -5.496812
2     4.0  0.231002 -2.886750
3     7.0  0.200275 -3.534375
4     3.0  0.291936 -4.352062

## Data Processing To Run It for 1 And 5

In [14]:
print(data[data['result']==1].shape[0])
print(data[data['result']==5].shape[0])

1005
556


In [15]:
data1=data[data['result']==1]
data5=data[data['result']==5]
data1=data1.append(data5)
data=data1
data.reset_index(drop=True,inplace=True)

In [16]:
data.head()

result  feature1  feature2
0     1.0  0.123043 -0.707875
1     1.0  0.113859 -0.931375
2     1.0  0.115371 -0.386000
3     1.0  0.102281 -0.378812
4     1.0  0.142043 -1.819438

In [17]:
data1=test_data[test_data['result']==1]
data5=test_data[test_data['result']==5]
data1=data1.append(data5)
test_data=data1
test_data.reset_index(drop=True,inplace=True)

In [18]:
test_data.head()

result  feature1  feature2
0     1.0  0.159934 -1.742000
1     1.0  0.124609 -0.474438
2     1.0  0.219117 -0.904500
3     1.0  0.244779 -2.505625
4     1.0  0.105563 -0.218562

# Question 4.a
Consider the linear kernel given in assignment. Train using the provided training data
and test using the provided test data, and report your accuracy over the entire test set,
and the number of support vectors.

In [19]:
from sklearn import svm

In [20]:
lin_clf=svm.SVC(kernel='linear')
lin_clf.kernel

'linear'

In [21]:
X=data[['feature1','feature2']]
Y=data['result']

In [22]:
lin_clf.fit(X,Y)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [23]:
test_data1=test_data
test_data1['predicted_output']=lin_clf.predict(test_data[['feature1','feature2']])

In [24]:
test_data1.head()

result  feature1  feature2  predicted_output
0     1.0  0.159934 -1.742000               1.0
1     1.0  0.124609 -0.474438               1.0
2     1.0  0.219117 -0.904500               1.0
3     1.0  0.244779 -2.505625               5.0
4     1.0  0.105563 -0.218562               1.0

In [25]:
accuracy=np.sum(test_data1['predicted_output']==test_data1['result'])/test_data1.shape[0]
print("Accuracy over entire test set:",accuracy)
print("Total support vectors (for each class resp.):",lin_clf.n_support_)

Accuracy over entire test set: 0.9787735849056604
Total support vectors (for each class resp.): [14 14]


# Question 4.b

In continuation, train only using the first {50, 100, 200, 800} points with the linear
kernel. Report the accuracy over the entire test set, and the number of support vectors
in each of these cases.

In [26]:
data1=data.sample(frac=1)
data1.head()

result  feature1  feature2
868     1.0  0.134496 -0.810813
417     1.0  0.209145 -0.481812
434     1.0  0.139037 -0.868563
754     1.0  0.146313 -0.552937
251     1.0  0.115732 -0.406813

In [27]:
size_list={50,100,200,800}
accuracy_list=[]
support_vectors=[]
Dict1 = {} #to store number of points and accuracy pair
Dict2 = {} #to store number of points and number of support vectors pair
for i in size_list:
  #Assigning Size of the Trainig Data
  train_data=data1[:i]
  #linear Classifier
  lin_clf=svm.SVC(kernel='linear')
  #preparing Data To Give Input to the Algorithm
  X=train_data[['feature1','feature2']]
  Y=train_data['result']
  #Training The Model
  lin_clf.fit(X,Y)
  test_data1=test_data
  test_data1['predicted_output']=lin_clf.predict(test_data1[['feature1','feature2']])
  accuracy=np.sum(test_data1['predicted_output']==test_data1['result'])/test_data1.shape[0]
  accuracy_list.append(accuracy)
  Dict1.update( {i:accuracy} )
  #Appending #Support Vectors
  s=lin_clf.n_support_
  Dict2.update({i:s[0]+s[1]})
print("Number of points and accuracy are given as:\n",Dict1) 
print("Number of points and Number of support vectors are given as:\n",Dict2) 

Number of points and accuracy are given as:
 {200: 0.9834905660377359, 50: 0.9811320754716981, 800: 0.9787735849056604, 100: 0.9811320754716981}
Number of points and Number of support vectors are given as:
 {200: 4, 50: 4, 800: 16, 100: 4}


# Question 4.c

Consider the polynomial kernel K(xn, xm) = (1 + x
T
n xm)
Q, where Q is the degree of
the polynomial. Comparing Q = 2 with Q = 5,
comment whether each of the following
case is TRUE or FALSE.

In [28]:
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.model_selection import cross_val_score

## When C=0.0001 training Error

In [62]:
poly_clf_2=svm.SVC(kernel='poly',degree=2,C=0.0001,coef0=1)
poly_clf_5=svm.SVC(kernel='poly',degree=5,C=0.0001,coef0=1)

X=data[['feature1','feature2']]
Y=data['result']

poly_clf_2.fit(X,Y)
poly_clf_5.fit(X,Y)

score_2=cross_val_score(poly_clf_2,X,Y,cv=5)
score_5=cross_val_score(poly_clf_5,X,Y,cv=5)

print("Training error at Q=2:",1-score_2.mean(),"\nTraining error at Q=5:" ,1-score_5.mean())

Training error at Q=2: 0.3267141803883019 
Training error at Q=5: 0.01921643319406885


(i) Statement: When C = 0.0001, training error is higher at Q = 5.
->Training Error Is Higher at Q=2. Therefore, statement (i) is **false**.

### When C = 0.001, the number of support vectors .

In [66]:
poly_clf_2=svm.SVC(kernel='poly',degree=2,C=0.001,coef0=1)
poly_clf_5=svm.SVC(kernel='poly',degree=5,C=0.001,coef0=1)

X=data[['feature1','feature2']]
Y=data['result']

poly_clf_2.fit(X,Y)
poly_clf_5.fit(X,Y)

score_2=cross_val_score(poly_clf_2,X,Y,cv=5)
score_5=cross_val_score(poly_clf_5,X,Y,cv=5)

print("Number of Support Vectors For Q=2:", poly_clf_2.n_support_,"\n Number of Support Vectors For Q=5:",poly_clf_5.n_support_)

Number of Support Vectors For Q=2: [228 228] 
 Number of Support Vectors For Q=5: [36 36]


(ii) Statement: When C = 0.001, the number of support vectors is lower at Q = 5.
     ->As we can see number of support vectors is lower at Q = 5.     
     Therefore, statement (ii) is **true**.

### When C = 0.01, training error 

In [31]:
poly_clf_2=svm.SVC(kernel='poly',degree=2,C=0.01,coef0=1)
poly_clf_5=svm.SVC(kernel='poly',degree=5,C=0.01,coef0=1)

X=data[['feature1','feature2']]
Y=data['result']

poly_clf_2.fit(X,Y)
poly_clf_5.fit(X,Y)

score_2=cross_val_score(poly_clf_2,X,Y,cv=5)
score_5=cross_val_score(poly_clf_5,X,Y,cv=5)

print("Training error at Q=2:",1-score_2.mean(),"\nTraining error at Q=5:" ,1-score_5.mean())

Training error at Q=2: 0.00640411239452765 
Training error at Q=5: 0.004485131481936522


(iii) Statement: When C = 0.01, training error is higher at Q = 5.
      ->As we can see training error is lower at Q = 5.
      Therefore, statement (iii) is **false**

### When C = 1, test error:

In [65]:
poly_clf_2=svm.SVC(kernel='poly',degree=2,C=1,coef0=1)
poly_clf_5=svm.SVC(kernel='poly',degree=5,C=1,coef0=1)

X=data[['feature1','feature2']]
Y=data['result']

poly_clf_2.fit(X,Y)
poly_clf_5.fit(X,Y)

score_2=cross_val_score(poly_clf_2,X,Y,cv=5)
score_5=cross_val_score(poly_clf_5,X,Y,cv=5)

test_data2=test_data

test_data2['predicted_output_2']=poly_clf_2.predict(test_data2[['feature1','feature2']])
accuracy_2=np.sum(test_data2['predicted_output_2']==test_data2['result'])/test_data2.shape[0]

test_data2['predicted_output_5']=poly_clf_5.predict(test_data2[['feature1','feature2']])
accuracy_5=np.sum(test_data2['predicted_output_5']==test_data2['result'])/test_data2.shape[0]

print("Test error at Q=2: ",1-accuracy_2,"\nTest error at Q=5: ",1-accuracy_5)

Test error at Q=2:  0.021226415094339646 
Test error at Q=5:  0.02358490566037741


(iv) Statement: When C = 1, test error is lower at Q = 5.
     ->As we can see error at Q = 5 is slightly higher than error at Q = 2 
     Therefore, above statement is **false**

# Question 4.d

Consider the radial basis function (RBF) kernel K(xn, xm) finding Training and testing error for different value of C

In [33]:
from sklearn import metrics

In [34]:
C =[0.01, 1, 100, 10**4,10**6]
training_error=[]
test_error=[]
for i in C:
  rbf_clf=svm.SVC(kernel='rbf',C=i,gamma=1)

  X=data[['feature1','feature2']]
  Y=data['result']

  rbf_clf.fit(X,Y)
  train_err=1-cross_val_score(rbf_clf,X,Y,cv=5)

  training_error.append(train_err.mean())


  pred=rbf_clf.predict(test_data[['feature1','feature2']])

  test_error.append(1-metrics.accuracy_score(test_data['result'],pred))




print("Training error is Minimum at C:= ",C[np.argmin(training_error)])
print("Test error is Minimum at C:= ",C[np.argmin(test_error)])

print("Training error values for c= 0.01,1,100,10^4,10^6 respectively are: ",training_error)
print("Test error values for c= 0.01,1,100,10^4,10^6 respectively are:",test_error)

Training error is Minimum at C:=  0.01
Test error is Minimum at C:=  100
Training error values for c= 0.01,1,100,10^4,10^6 respectively are:  [0.005126157122962227, 0.006408208405013505, 0.005126157122962227, 0.01089538789219302, 0.01089538789219302]
Test error values for c= 0.01,1,100,10^4,10^6 respectively are: [0.02358490566037741, 0.021226415094339646, 0.018867924528301883, 0.02358490566037741, 0.02358490566037741]


# Assignment Question 5


In [35]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_train.labels

--2021-10-07 16:01:18--  https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_train.labels
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 15000 (15K) [application/x-httpd-php]
Saving to: ‘gisette_train.labels.2’

gisette_train.label 100%[===================>]  14.65K  --.-KB/s    in 0.06s   

2021-10-07 16:01:18 (232 KB/s) - ‘gisette_train.labels.2’ saved [15000/15000]



In [36]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/gisette_valid.labels

--2021-10-07 16:01:20--  https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/gisette_valid.labels
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2500 (2.4K) [application/x-httpd-php]
Saving to: ‘gisette_valid.labels.2’

gisette_valid.label 100%[===================>]   2.44K  --.-KB/s    in 0s      

2021-10-07 16:01:21 (81.1 MB/s) - ‘gisette_valid.labels.2’ saved [2500/2500]



In [37]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_train.data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_valid.data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_test.data

--2021-10-07 16:01:22--  https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_train.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 67788724 (65M) [application/x-httpd-php]
Saving to: ‘gisette_train.data.2’

gisette_train.data. 100%[===================>]  64.65M  34.1MB/s    in 1.9s    

2021-10-07 16:01:25 (34.1 MB/s) - ‘gisette_train.data.2’ saved [67788724/67788724]

--2021-10-07 16:01:25--  https://archive.ics.uci.edu/ml/machine-learning-databases/gisette/GISETTE/gisette_valid.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 11316914 (11M) [application/x-httpd-php]
Saving to: ‘gisette_valid.data.2’

gisette_valid.dat

In [38]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn import metrics

In [39]:
header_list=np.arange(5000)
train_data=pd.read_csv("/content/gisette_train.data",sep='\s+',names=header_list)
train_data_y=pd.read_csv("/content/gisette_train.labels",names=['result'])
valid_data=pd.read_csv("/content/gisette_valid.data",sep='\s+',names=header_list)
valid_data_y=pd.read_csv("/content/gisette_valid.labels",names=['result'])

In [40]:
train_data.head()

0     1     2     3     4     5     ...  4994  4995  4996  4997  4998  4999
0   550     0   495     0     0     0  ...   991     0     0     0     0   983
1     0     0     0     0     0     0  ...     0     0   991     0     0     0
2     0     0     0     0     0     0  ...     0     0     0     0     0     0
3     0     0   742     0     0     0  ...     0     0   674     0     0   838
4     0     0     0     0     0     0  ...     0     0     0     0     0     0

[5 rows x 5000 columns]

In [41]:
train_data_y.head()

result
0       1
1      -1
2       1
3       1
4       1

In [42]:
valid_data.head()

0     1     2     3     4     5     ...  4994  4995  4996  4997  4998  4999
0   688     0     0     0     0     0  ...     0     0   494     0   769     0
1   778   758     0     0     0     0  ...     0     0     0     0     0     0
2   469     0   816     0     0     0  ...     0     0     0   417     0     0
3     0     0     0     0     0   571  ...     0     0     0     0     0     0
4     0     0     0     0     0     0  ...     0     0     0     0   976     0

[5 rows x 5000 columns]

In [43]:
valid_data_y.head()

result
0       1
1       1
2      -1
3       1
4       1

In [44]:
print(valid_data.shape,valid_data_y.shape)

(1000, 5000) (1000, 1)


In [45]:
print(train_data.shape)

(6000, 5000)


## Question 5.a

Training The Data Using Linear Kernal

In [46]:
lin=svm.SVC(kernel='linear')


In [47]:
print(train_data.shape,train_data_y.shape)

(6000, 5000) (6000, 1)


In [48]:
lin.fit(train_data,train_data_y['result'])

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [67]:
print("Total support vectors :",lin.n_support_)

Total support vectors : [542 542]


Therefore total support vectors are 542 for each class. Which means total support vectors are **1084**

#### Training Error

In [50]:
train_err=1-cross_val_score(lin,train_data,train_data_y['result'],cv=5)
print("Training error :",train_err.mean())

Training error : 0.02733333333333332


Therefore, Training error is **0.0273**

In [51]:
valid_data_y['predicted']=lin.predict(valid_data)
test_acc=metrics.accuracy_score(valid_data_y['result'],valid_data_y['predicted'])
acc=np.sum(valid_data_y['result']==valid_data_y['predicted'])/valid_data_y.shape[0]
print("Test error:",1-acc)

Test error: 0.02400000000000002


Therefore, Testing error is **0.024**

In [52]:
valid_data_y

result  predicted
0         1         -1
1         1          1
2        -1         -1
3         1          1
4         1          1
..      ...        ...
995      -1         -1
996       1          1
997       1          1
998      -1         -1
999      -1         -1

[1000 rows x 2 columns]

# Question 5.b

Using RBF and Polynomial Kernel

1) RBF Kernel

In [53]:
rbf=svm.SVC?

In [54]:
rbf=svm.SVC(kernel='rbf',gamma=0.001)

In [55]:
rbf.fit(train_data,train_data_y['result'])

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.001, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [57]:
train_err=1-cross_val_score(rbf,train_data,train_data_y['result'],cv=5)
print("Training error: ",train_err.mean())
valid_data_y['predicted_rbf']=rbf.predict(valid_data)
test_accuracy=metrics.accuracy_score(valid_data_y['result'],valid_data_y['predicted_rbf'])
acc=np.sum(valid_data_y['result']==valid_data_y['predicted_rbf'])/valid_data_y.shape[0]
print("Testing error: ",1-acc)
print("Number Of Support Vectors: ",rbf.n_support_)

Training error:  0.5
Testing error:  0.5
Number Of Support Vectors:  [3000 3000]


In [58]:
valid_data_y.head()

result  predicted  predicted_rbf
0       1         -1              1
1       1          1              1
2      -1         -1              1
3       1          1              1
4       1          1              1

In [59]:
poly=svm.SVC(kernel='poly',coef0=0.001,degree=2)

In [61]:
poly.fit(train_data,train_data_y['result'])

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.001,
    decision_function_shape='ovr', degree=2, gamma='scale', kernel='poly',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [64]:
train_err=1-cross_val_score(poly,train_data,train_data_y['result'],cv=5)
print("Training error: ",train_err.mean())
valid_data_y['predicted_poly']=poly.predict(valid_data)
test_acc=metrics.accuracy_score(valid_data_y['result'],valid_data_y['predicted_poly'])
acc=np.sum(valid_data_y['result']==valid_data_y['predicted_poly'])/valid_data_y.shape[0]
print("Testing error: ",1-acc)
print("Number Of Support Vectors: ",poly.n_support_)

Training error:  0.023833333333333328
Testing error:  0.020000000000000018
Number Of Support Vectors:  [820 937]
